src/Compilers/CSharp/Portable/Lowering/RuntimeAsyncIteratorRewriter/RuntimeAsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,428 @@
+
+        // this.state = {initialState};
+        // this.disposeMode = false;
+
+        builder.Add(
+            // this.state = {initialState};
+            F.Assignment(F.Field(F.This(), stateField), F.Literal(initialState)));
+
+        builder.Add(
+            // disposeMode = false;
+            F.Assignment(F.InstanceField(_disposeModeField), F.Literal(false)));
+    }
+
+    protected override BoundStatement InitializeParameterField(MethodSymbol getEnumeratorMethod, ParameterSymbol parameter, BoundExpression resultParameter, BoundExpression parameterProxy)
+    {
+        return AsyncRewriter.AsyncIteratorRewriter.InitializeParameterField(getEnumeratorMethod, parameter, resultParameter, parameterProxy, _combinedTokensField, F);
+    }
+
+    /// <summary>
+    /// Generates the `ValueTask&gt;bool> IAsyncEnumerator&gt;ElementType>.MoveNextAsync()` method as a runtime-async method.
+    /// </summary>
+    [SuppressMessage("Style", """VSTHRD200:Use "Async" suffix for async methods""", Justification = "Standard naming convention for generating 'IAsyncEnumerator.MoveNextAsync'")]
+    private void GenerateIAsyncEnumeratorImplementation_MoveNextAsync()
+    {
+        Debug.Assert(stateField is not null);
+        Debug.Assert(_currentField is not null);
+        Debug.Assert(_disposeModeField is not null);
+        Debug.Assert(hoistedVariables is not null);
+        Debug.Assert(nonReusableLocalProxies is not null);
+
+        // Add IAsyncEnumerator<...>.MoveNextAsync() as a runtime-async method.
+        MethodSymbol IAsyncEnumeratorOfElementType_MoveNextAsync = GetMoveNextAsyncMethod();
+        SynthesizedImplementationMethod moveNextAsyncMethod = OpenMoveNextMethodImplementation(IAsyncEnumeratorOfElementType_MoveNextAsync, runtimeAsync: true);
+
+        var rewritter = new MoveNextAsyncRewriter(
+             F,
+             method,
+             stateField,
+             instanceIdField,
+             hoistedVariables,
+             nonReusableLocalProxies,
+             nonReusableFieldsForCleanup,
+             synthesizedLocalOrdinals,
+             stateMachineStateDebugInfoBuilder,
+             slotAllocatorOpt,
+             nextFreeHoistedLocalSlot,
+             diagnostics,
+             _currentField,
+             _disposeModeField,
+             _combinedTokensField);
+
+        BoundStatement runtimeAsyncMoveNextAsyncBody = rewritter.GenerateMoveNextAsync(body);
+
+        Debug.Assert(F.CurrentFunction is not null);
+        BoundStatement rewrittenBody = RuntimeAsyncRewriter.RewriteWithoutHoisting(runtimeAsyncMoveNextAsyncBody, F.CurrentFunction, F.CompilationState, F.Diagnostics);
+
+        F.CloseMethod(rewrittenBody);
+    }
+
+    private MethodSymbol GetMoveNextAsyncMethod()
+    {
+        Debug.Assert(_currentField is not null);
+
+        MethodSymbol IAsyncEnumerator_MoveNextAsync = F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync);
```

Comment:
Please assert that the return type of this method is `ValueTask<T>` from the BCL, not another source.